docs/RFCS/20200331_enums.md

```diff
@@ -0,0 +1,298 @@
+arrays to create the ordering that we need. In particular we reserve
+the minimum byte (`0`) and have a maximum allowed byte. In practice
+this will be `255`. An example of the encoding scheme is below.
+
+Assume we started with 3 elements (`a`, `b`, `c`), and let the maximum byte value be 3.
+The sort order byte arrays for each element would be:
+```
+a 1/
+b 2/
+c 3/
+```
+To add an element after `b` we can create a new key that sits in the middle of the range
+between `b` and `c`.
+```
+a 1/
+b 2/
+d 2/2/
+c 3/
+```
+Now lets add more values before `d`. The first one is easy:
+```
+a 1/
+b 2/
+e 2/1/
+d 2/2/
+c 3/
+```
+The tricky case is adding a value before `e`. Because we reserved the minimum byte, we can
+append it and then bisect the range again.
+```
+a 1/
+b 2/
+f 2/0/2
+e 2/1/
+d 2/2/
+c 3/
+```
+This strategy can be extended indefinitely as long as this pattern is followed to reserve
+the minimum byte. A prototype of the exact algorithm is included as part of the RFC PR.
+
+This sort order byte array will be the physical layout and identifier of the enum. We expect
+that for small enums only a byte or two will be used to hold all the values, and that our
+compression strategies at the storage layer will compress this data well.
+
+An additional strategy is to represent enum values as a tuple of an integer and 
+this byte array.
+The ordering of the enums will be first on the integer, then on the byte array. An advantage
+of this approach is that adding a new enum to the beginning or end of the set requires only
+an increment/decrement of the integer rather than creating a new and possibly longer byte array.
+Adding to the beginning or end of the existing enum value set is the most
+common operation, and is the "worst case" for the pure byte array algorithm.
+A drawback of this strategy is that more space is used in the case where there are not many enum
+values. 
+
+## Parsing Changes
+
+Currently, the CockroachDB grammar is not equipped to handle type names
+that are qualified due to changes made in the past that separated parsing of
+object and type identifiers. Some of these changes will have to be 
+reverted/adapted in order to allow for types to have qualifications again.
+I am currently investigating the extend of the changes that need to be made in
+this area. An example of these changes are in 
+[this PR](https://github.com/cockroachdb/cockroach/pull/47216).
+
+
+## Type System Changes
+
+To update the type system to handle enums, we will introduce a new enum type
+which holds onto the ID of the enum. Using this, only enums of the same type
+can be compared etc. A new `Datum` `DEnum` will be introduced that holds 
+onto the ID of the enum that it represents as well as the physical layout 
+discussed above. Operations on two `DEnum` datums will succeed only if the
+two `DEnum`s have the same enum ID.
+
+The type system itself will also have to be extended to have support for
```

Comment:
I think we need more detail here. We have an existing `types.proto` file. I'd like to understand how that would be updated to handle enums, and eventually arbitrary PG user defined types. Today we can represent fairly rich structured types (tuple, array, simple types). Tuples even support named fields. I think the main thing we're lacking is a way to give the top-level type a user-defined name.

For this reason, my first instinct would be that we need to extend the `InternalType` proto (in `types.proto`) with some UDT-related fields + versioning field(s). Could you explain why we'd need a separate `UserDefinedTypeDescriptor` class?

Reply:
The main reason is that the types proto is stored on disk in different data structures, like table descriptors. Some user defined type data (like the name, or the field labels in an enum) needs to be changeable without having to update all existing descriptors.

Reply:
These descriptors need to also store metadata that is really unrelated to the types themselves, like what tables and columns use the type, and the parent schema and database the type resides in.